Tests for colocation_auto

ejgal opened the pull request:
working on #132

ejgal commented:
@jgliss: Not sure if the fixtures I define here are the best, let me know if you have a better idea.
All tests from line 136 and down are for functions that are not run in the class or from any other place in pyaerocom as far as I can see.
Should I create an issue on that and trim the class down a bit, or are there plans for using those for something?

jgliss commented on pyaerocom/test/test_colocation_auto.py:
```diff
@@ -0,0 +1,181 @@
+@pytest.fixture(scope='module')
+def init_obsdata_paths():
+    import pyaerocom as pya
+    pya.const.add_data_search_dir(TESTDATADIR.joinpath('obsdata'))
```
What is the advantage of having this as a fixture, instead of just adding this search directory in conftest?

ejgal replied:
None that I can think of, will move it!

jgliss commented on pyaerocom/test/test_colocation_auto.py:
```diff
@@ -0,0 +1,181 @@
+    import pyaerocom as pya
+    pya.const.add_data_search_dir(TESTDATADIR.joinpath('obsdata'))
+
+@pytest.fixture(scope='function')
+def col_tm5_aero(model_id='TM5-met2010_CTRL-TEST', obs_id='AeronetSunV3L2Subset.daily',
+                 obs_vars='od550aer'):
+    coloc = Colocator(model_id=model_id, obs_id=obs_id, obs_vars=obs_vars)
+    coloc.raise_exceptions = True
+    coloc.reanalyse_existing = True
+    coloc.start = 2010
+    return coloc
```
Using `scope='function'` seems to be an interesting concept that I have not used before. I trust you that this makes sense, as I do not have the time right now to really understand why you do it this way.

ejgal replied:
This is done since some of the tests might change the properties of coloc.
Should be okay since the fixture is not doing anything resource intensive.

jgliss commented on pyaerocom/test/test_colocation_auto.py:
```diff
@@ -0,0 +1,181 @@
+    col.update(basedir_coldata=basedir)
+    assert os.path.isdir(basedir)
+
+def test_colocator_run(init_obsdata_paths, col_tm5_aero):
+    # gridded_ungridded
+    col_tm5_aero.run()
+    coldata = col_tm5_aero.data[col_tm5_aero.model_id][col_tm5_aero.obs_vars[0]]
+    assert isinstance(coldata, ColocatedData)
+
+    # gridded_gridded
+    col_tm5_aero.obs_id = col_tm5_aero.model_id
+    col_tm5_aero.run()
+    coldata = col_tm5_aero.data[col_tm5_aero.model_id][col_tm5_aero.obs_vars[0]]
+    assert isinstance(coldata, ColocatedData)
```
2 tests in 1? parametrize?

jgliss commented on pyaerocom/test/test_colocation_auto.py:
```diff
@@ -0,0 +1,181 @@
+def test_colocator__find_var_matches_OLD(): # Never called, delete?
+    pass
+
+def test_colocator_output_dir(): # Never called, delete?
+    col = Colocator(raise_exceptions=True)
+    with pytest.raises(AttributeError):
+        col.output_dir('task')
```
Seems like method `output_dir` can indeed be deleted from Colocator class. Does not work anyways :)

jgliss commented on pyaerocom/test/test_colocation_auto.py:
```diff
@@ -0,0 +1,181 @@
+    with pytest.raises(NotImplementedError):
+        col()
+
+def test_colocator__find_var_matches_OLD(): # Never called, delete?
+    pass
```
Feel free to remove this from Colocator class

jgliss commented on pyaerocom/test/test_colocation_auto.py:
```diff
@@ -0,0 +1,181 @@
+    data = col.read_model_data('od550aer')
+    assert isinstance(data, GriddedData)
+
+def test_colocator_call(): # Delete __call__?
+    col = Colocator(raise_exceptions=True)
+    with pytest.raises(NotImplementedError):
+        col()
```
Maybe we want to re-introduce __call__ ?

jgliss replied:
Maybe we want to re-introduce method `__call__` ?

jgliss commented on pyaerocom/test/test_colocation_auto.py:
```diff
@@ -0,0 +1,181 @@
+
+### Remove?
+
+def test_colocator_read_ungridded(): # Never called, delete?
+    col = Colocator(raise_exceptions=True)
+    obs_id = 'AeronetSunV3L2Subset.daily'
+    obs_var = 'od550aer'
+    col.obs_filters = {'longitude' : [-30, 30]}
+    col.obs_id = obs_id
+    col.read_opts_ungridded = {'last_file' : 10}
+    with pytest.raises(DataCoverageError):
+        data = col.read_ungridded(obs_var) # Why is this raised?
+        # read_ungridded expects a list of one or more variables!
+        # Should check for str and convert to list?
+        # Should read_ungridded and read_model_data have same arguments?
+    data = col.read_ungridded([obs_var])
+    assert isinstance(data, UngriddedData)
+
+    col.read_opts_ungridded = None
+    col.obs_vars = ['od550aer']
+    with does_not_raise_exception():
+        data = col.read_ungridded()
+    col.obs_vars = ['invalid']
+    with pytest.raises(DataCoverageError):
+        data = col.read_ungridded()
```
I would like to keep this method in `Colocator` class as I use it in some scripts for quick checks I think

jgliss commented on pyaerocom/test/test_colocation_auto.py:
```diff
@@ -0,0 +1,181 @@
+    with pytest.raises(DataCoverageError):
+        data = col.read_ungridded()
+
+def test_colocator_read_model_data(): # Never called, delete?
+    col = Colocator(raise_exceptions=True)
+    model_id = 'TM5-met2010_CTRL-TEST'
+    col.model_id = model_id
+    data = col.read_model_data('od550aer')
+    assert isinstance(data, GriddedData)
```
Please keep, need this method in some of my scripts.

jgliss replied:
I mean the actual method in `Colocator` of course, not the test method :) But having a test for this is great as well!

ejgal commented:
Parametrized test_colocator_run and changed the obsdata fixture to just be a path in TEST_DATA.
Also removed the unnecessary functions and kept the others.
